nemoguardrails/library/autoguard/actions.py

```diff
@@ -0,0 +1,496 @@
+            response_json = []
+            for line in data.split("\n"):
+                line = line.strip()
+                if len(line) > 0:
+                    response_json.append(json.loads(line))
+            log.info(json.dumps(response_json, indent=True))
+            print(response_json)
+            for i in response_json:
+                if i["guarded"]:
+                    return True
+            return False
+
+
+@action(name="call autoguard race bias api", is_system_action=True)
+async def call_autoguard_race_bias_api(context: Optional[dict] = None):
+    api_key = os.environ.get("AUTOGUARD_API_KEY")
+
+    if api_key is None:
+        raise ValueError("AUTOGUARD_API_KEY environment variable not set.")
+
+    user_message = context.get("user_message")
+    bot_message = context.get("bot_message")
+    if bot_message:
+        prompt = bot_message
+    else:
+        prompt = user_message
+
+    model = context.get("model")
+    url = "http://35.225.99.81:8888/"
+
+    if model:
+        url = url + "query"
+    else:
+        url = url + "guardrail"
+
+    headers = {"x-api-key": api_key}
+
+    data = {
+        "config": {
+            "gender_bias_detection": {"mode": "OFF"},
+            "tonal_detection": {"mode": "OFF"},
+            "pii_fast": {
+                "mode": "OFF",
+                "enabled_types": [],
+                "mask": False,
+                "coreference": False,
+            },
+            "factcheck": {"mode": "OFF"},
+            "confidential_detection": {"mode": "OFF"},
+            "jailbreak_detection": {"mode": "OFF"},
+            "text_toxicity_extraction": {"mode": "OFF"},
+            "harm_detection": {"mode": "OFF"},
+            "intellectual_property": {"mode": "OFF"},
+            "racial_bias_detection": {"mode": "DETECT"},
+        },
+        "prompt": prompt,
+    }
+    if model:
+        data["model"] = model
+    async with aiohttp.ClientSession() as session:
+        async with session.post(
+            url=url,
+            headers=headers,
+            json=data,
+        ) as response:
+            if response.status != 200:
+                raise ValueError(
+                    f"AutoGuard call failed with status code {response.status}.\n"
+                    f"Details: {await response.text()}"
+                )
+            data = await response.text()
+            response_json = []
+            for line in data.split("\n"):
+                line = line.strip()
+                if len(line) > 0:
+                    response_json.append(json.loads(line))
+            log.info(json.dumps(response_json, indent=True))
+            print(prompt)
+            print(response_json)
+            for i in response_json:
+                if i["guarded"]:
+                    return True
+            return False
+
+
+@action(name="call autoguard harm detection api", is_system_action=True)
+async def call_autoguard_harm_detection_api(context: Optional[dict] = None):
```

Comment:
all these functions seem to have lots of code in common. instead of repeating them, can we use some sort of Template Design Pattern?https://www.geeksforgeeks.org/template-method-design-pattern/

Reply:
resolved in previous commits